ClusterTrustBundles - move to beta

stlaz opened the pull request:
#### What type of PR is this?
/kind api-change
/kind feature

#### What this PR does / why we need it:
This PR bumps the ClusterTrustBundle API to v1beta1. It also adds API discovery for kubelet to determine whether to enable the ClusterTrustBundleProjection feature based on the API availability (the feature gate must still be enabled for this to happen).

#### Which issue(s) this PR fixes:
Related to: https://github.com/kubernetes/enhancements/issues/3257

#### Special notes for your reviewer:
This PR is rebased on https://github.com/kubernetes/kubernetes/pull/127326 and should not merge before the other PR.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The ClusterTrustBundle API is moving to v1beta1.
In order for the ClusterTrustBundleProjection feature to work on the kubelet side, the ClusterTrustBundle API must be available at v1beta1 version and the ClusterTrustBundleProjection feature gate must be enabled. If the API becomes later after kubelet started running, restart the kubelet to enable the feature.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3257
```

stlaz committed:
copy ClusterTrustBundle API to v1beta1

stlaz committed:
use the ClusterTrustBundles beta API

stlaz committed:
kubelet: lazy enabling the ClusterTrustBundleProjection feature

Determine whether the ClusterTrustBundleProjection should be enabled
based on ClusterTrustBundle API discovery.
Some distributions may rely on a running kubelet in order to start
their kube-apiserver. Therefore we must delay the API discovery.

This patch delays it until the first time a clustertrustbundle is
requested from the InformerMaanager.

stlaz committed:
KCM: CTBPublisher: use generics to handle both alpha/beta APIs

stlaz committed:
kubelet: ctb: use generics to handle alpha/beta APIs for CTB projection

stlaz committed:
integration: ctb: apiserversigner test to use certbeta api

enj reviewed (changes requested):
Quick pass on some of the files.

ahmedtd commented on pkg/apis/certificates/v1beta1/conversion.go:
```diff
@@ -24,15 +24,21 @@ import (
-		},
-	)
+	for _, kind := range []string{"CertificateSigningRequest", "ClusterTrustBundle"} {
+		err := scheme.AddFieldLabelConversionFunc(SchemeGroupVersion.WithKind(kind),
```
Nit: I think this sharing is a little too clever --- CertificateSigningRequest and ClusterTrustBundle are separate types, and the fact that they currently have identical field->label conversion funcs is just accidental.

Can we just write out both definitions separately?  Otherwise someone who evolves this in the future may be tempted to start adding `if`s inside the loop.

stlaz replied:
I originally had it written exactly that way but then I thought that future us must be better than to just shove an additional `if` in there, and so I switched to the loop logic :smile: 

I'll move it back, I think it makes sense to treat different objects differently :slightly_smiling_face:

ahmedtd commented on pkg/controller/certificates/clustertrustbundlepublisher/publisher.go:
```diff
@@ -0,0 +1,365 @@
+
+// clusterTrustBundle is a type constraint grouping all APIs versions of ClusterTrustBundles
+type clusterTrustBundle interface {
+	certificatesv1alpha1.ClusterTrustBundle | certificatesv1beta1.ClusterTrustBundle
```
I don't understand why this is necessary --- why should this controller need to handle v1alpha1 ClusterTrustBundles?

stlaz replied:
@enj thought it might be better to work with either API version

enj replied:
We will definitely need it for beta -> v1 so no reason not do it here as well.

ahmedtd commented on pkg/kubelet/clustertrustbundle/clustertrustbundle_manager.go:
```diff
@@ -40,6 +46,42 @@ const (
+	GetTrustBundle(*T) string
+}
+
+type alphaClusterTrustBundleHandlers struct{}
```
Same question about handling alpha APIs here... isn't it fine to just straight-up break people who were using the alpha API?

fedebongio commented:
/remove-sig api-machinery

seans3 commented:
/remove-label api-machinery

seans3 commented:
/remove-sig api-machinery

stlaz commented:
Improved the manager's lazy init and added unit tests for it.

Jefftree commented:
/remove-label api-machinery

neolit123 reviewed (commented):
/remove-area kubeadm
/remove-sig cluster-lifecycle

stlaz commented:
Resolved the merge conflict in the e2e test.

stlaz commented:
The test failures observed on the last rebase didn't seem to be caused by this PR, rebasing again to see if this is fixed in master.
The `apidiff-client-go` seems misconfigured as, as far as I understand the description, it marked modules
```
- ./informers/certificates/v1beta1.Interface.ClusterTrustBundles: added
- ./kubernetes/typed/certificates/v1beta1.ClusterTrustBundlesGetter.ClusterTrustBundles: added 
```
as internal.

stlaz committed:
move ClusterTrustBundles featuregates to beta

stlaz committed:
generate code

stlaz commented:
When would usually be the right time to change the spec of `-e2e-gce-cos-alpha-features` not to ginkgo.focus the tests that are no longer available in beta?

fedebongio commented:
/remove-sig api-machinery

stlaz commented on pkg/printers/internalversion/printers.go:
```diff
@@ -31,7 +31,7 @@ import (
 	batchv1 "k8s.io/api/batch/v1"
 	batchv1beta1 "k8s.io/api/batch/v1beta1"
-	certificatesv1alpha1 "k8s.io/api/certificates/v1alpha1"
+	certificatesv1alpha1 "k8s.io/api/certificates/v1alpha1" // should this change, too? there are still certv1beta1.CSR printers, but not their v1 versions
```
TODO for the review mtg to discuss

liggitt commented on cmd/kube-controller-manager/app/certificates.go:
```diff
@@ -243,36 +247,50 @@ func newKubeAPIServerSignerClusterTrustBundledPublisherController(ctx context.Co
+	var runner ctbpublisher.PublisherRunner
+	for _, gv := range []schema.GroupVersion{certificatesv1beta1.SchemeGroupVersion, certificatesv1alpha1.SchemeGroupVersion} {
+		ctbAvailable, err := clusterTrustBundlesAvailable(apiserverSignerClient, gv)
+		if err != nil {
```
I *think* we may need to tolerate NotFound errors (here or in clusterTrustBundlesAvailable) if we want to continue if the group/version is completely unserved (see if we have a way to test that as well)

liggitt commented on pkg/kubelet/kubelet.go:
```diff
@@ -875,18 +875,10 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
-		}
-		kubeInformers.Start(wait.NeverStop)
-		klog.InfoS("Started ClusterTrustBundle informer")
+		clusterTrustBundleManager = clustertrustbundle.NewLazyInformerManager(ctx, kubeDeps.KubeClient, 2*int(kubeCfg.MaxPods))
```
nit, might make sense to log something here that we're setting up this client

liggitt commented on pkg/kubelet/kubelet.go:
```diff
@@ -875,18 +875,10 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 	} else {
-		// In static kubelet mode, use a no-op manager.
-		clusterTrustBundleManager = &clustertrustbundle.NoopManager{}
 		klog.InfoS("Not starting ClusterTrustBundle informer because we are in static kubelet mode")
```
nit: this log message isn't quite accurate if KubeClient != nil but the feature is disabled

liggitt commented on pkg/kubelet/clustertrustbundle/clustertrustbundle_manager.go:
```diff
@@ -261,3 +323,105 @@ func (m *NoopManager) GetTrustAnchorsByName(name string, allowMissing bool) ([]b
+}
+
+func (m *LazyInformerManager) GetTrustAnchorsByName(name string, allowMissing bool) ([]byte, error) {
+	if m.manager == nil {
```
we have to protect all reads ... if we want to avoid contention after init, use a RWMutex

```
if err := m.ensureInitialized(); err != nil {
  return nil, err
}

func ensureInitialized() error {
RLock()
defer RUnlock()
check for nil, if non nil, return

Lock()
defer Unlock()
re-check for nil and init
}
```

liggitt commented on pkg/kubelet/clustertrustbundle/clustertrustbundle_manager.go:
```diff
@@ -261,3 +323,105 @@ func (m *NoopManager) GetTrustAnchorsByName(name string, allowMissing bool) ([]b
+
+	var clusterTrustBundleManager Manager
+	for _, gv := range []schema.GroupVersion{certificatesv1beta1.SchemeGroupVersion, certificatesv1alpha1.SchemeGroupVersion} {
+		ctbAPIAvailable, err := clusterTrustBundlesAvailable(m.client, gv)
```
similar comment about treating a NotFound on the group/version and a missing individual resource consistently

liggitt commented on pkg/kubelet/clustertrustbundle/clustertrustbundle_manager.go:
```diff
@@ -74,29 +136,29 @@ func NewInformerManager(ctx context.Context, bundles certinformersv1alpha1.Clust
 	_, err := m.ctbInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj any) {
-			ctb, ok := obj.(*certificatesv1alpha1.ClusterTrustBundle)
+			ctb, ok := obj.(*certificatesv1beta1.ClusterTrustBundle)
```
these probably have to be parameterized somehow... hard-coding beta types in the generic doesn't seem right (which also means we didn't test this working on the alpha)

stlaz committed:
KCM: CTB publisher: tolerate NotFound during discovery

liggitt commented:
apidiff flagging additions to existing interfaces, as expected:
```
Incompatible changes:
- ./kubernetes/typed/certificates/v1beta1.ClusterTrustBundlesGetter.ClusterTrustBundles: added
- ./informers/certificates/v1beta1.Interface.ClusterTrustBundles: added
```

no breaking removals or signature changes

liggitt commented on pkg/kubelet/clustertrustbundle/clustertrustbundle_manager.go:
```diff
@@ -261,3 +332,115 @@ func (m *NoopManager) GetTrustAnchorsByName(name string, allowMissing bool) ([]b
+	m.manager = clusterTrustBundleManager
+	kubeInformers.Start(m.contextWithLogger.Done())
+	m.logger.Info("Started ClusterTrustBundle informer")
+
```
At this point, we'll have a brand-new informer we just started, which will take a little while to sync in the happy path

If we instantly return here, the GetTrustAnchorsByName or GetTrustAnchorsBySigner call is likely to error with a "not yet synced" error. Will that propagate to something the kubelet will retry, or should we wait *briefly* here?

```go
	synced := true
	timeoutContext, cancel := context.WithTimeout(m.contextWithLogger, 10*time.Second)
	defer cancel()
	m.logger.Info("Waiting for ClusterTrustBundle informer to sync")
	for _, ok := range kubeInformers.WaitForCacheSync(timeoutContext.Done()) {
		synced = synced && ok
	}
	if synced {
		m.logger.Info("ClusterTrustBundle informer synced")
	} else {
		m.logger.Info("ClusterTrustBundle informer not synced, continuing to attempt in background")
	}
```

stlaz replied:
The error would eventually propagate all the way to https://github.com/kubernetes/kubernetes/blob/6f74f1f1bc5a5cd1f520827ffec164e006a858f0/pkg/volume/util/operationexecutor/operation_generator.go#L601-L607
The comment theres ays the operation would be retried.

However, I think it still makes sense to perform a short cache-sync wait given the cache fetch operation directly follows the setup - that's a very good point, I did not realize that the original code did not need this as there were likely more steps that followed the informer setup during kubelet start.

liggitt reviewed (commented):
code is looking good, a few last bits about logging and a question about waiting briefly for the initial informer sync

CI looks red... verify / codegen failures look like a regen is needed (I bet the generator changed)

integration failure looks spurious / pre-existing - https://storage.googleapis.com/k8s-triage/index.html?date=2025-02-13&pr=1&text=TestStorageVersionMigrationWithCRD:

```
    storageversionmigrator_test.go:215: CR not stored at version v2
--- FAIL: TestStorageVersionMigrationWithCRD (9.75s)
```
opened https://github.com/kubernetes/kubernetes/issues/130148

cici37 commented:
/remove-sig api-machinery

stlaz commented:
/triage accepted
this was triaged long ago but we did not label it

liggitt commented:
changes look good, it looks like the CTB e2e tests are failing in the pull-kubernetes-e2e-gce-cos-alpha-features job ... 

It looks like that job enables all APIs, but then only enables alpha features: `--env=KUBE_FEATURE_GATES=AllAlpha=true,EventedPLEG=false`

Might need to tweak that job to set ClusterTrustBundle and ClusterTrustBundleProjection to true explicitly to keep those features working as they shift to beta... mind doing that and seeing if we can get a clean run of that job before merge here?

stlaz commented:
I opened https://github.com/kubernetes/test-infra/pull/34451. It looks like there are more jobs that do the same, enabling all alpha APIs and focusing these tests.

stlaz commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features
the test should no longer fail as the CTBs are no longer a part of it

liggitt commented:
/test pull-kubernetes-cmd

liggitt commented:
/test pull-kubernetes-e2e-kind-alpha-features
(should pass and not include these tests)

/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-beta-features
(should pass and include these tests)

stlaz commented:
> /test pull-kubernetes-e2e-kind-alpha-features (should pass and not include these tests)
> 
> /test pull-kubernetes-e2e-kind-alpha-beta-features /test pull-kubernetes-e2e-kind-beta-features (should pass and include these tests)

Looking at the tests runs, this expectation seem to have been met.

liggitt commented:
/lgtm
/approve
/hold cancel

stlaz commented:
Looks like 56bd8d86a58d6060741ca8a8bd9873e77087d5d9 introduced changes that were explicitly adding some of the alpha API to tests in the meantime. This should now be fixed.

liggitt commented:
/lgtm
